### PR TITLE
system logs ui changes

### PR DESCRIPTION
**Pull request by vishalmishraa**
## Linked Issue(s)

issue #555 

## Acceptance Criteria fulfillment

- [ ] Task 1
- [ ] Task 2
- [ ] Task 3

## Proposed changes (including videos or screenshots)

Screenshot 2024-10-05 at 1 53 58 AM
Screenshot 2024-10-05 at 1 50 18 AM
Screenshot 2024-10-05 at 1 50 03 AM
Screenshot 2024-10-05 at 1 50 11 AM

### CODE REVIEW

**Comment by jayantbh** [comment]
Thanks for your contribution, @vishalmishraa!
Based on the screenshots it's certainly a decent improvement.

I would love to accept this PR, but it needs a few improvements to avoid performance issues.

Given that there could be thousands of lines being rendered, we do need to keep an eye out for excessive calls to JSON.stringify, Array.includes, and various other places.

Additionally, I think the metadata doesn't really need to be printed. What do you think?

**Comment by jayantbh** [comment]
Linter failed. We need to sort out the linting issues but for now you can check our contributing guide to see how to fix that. Let us know if it doesn't work out for you.

**Comment by vishalmishraa** [comment]
> Thanks for your contribution, @vishalmishraa! Based on the screenshots it's certainly a decent improvement.
> 
> I would love to accept this PR, but it needs a few improvements to avoid performance issues.
> 
> Given that there could be thousands of lines being rendered, we do need to keep an eye out for excessive calls to JSON.stringify, Array.includes, and various other places.
> 
> Additionally, I think the metadata doesn't really need to be printed. What do you think?

hi , @jayantbh , 

I removed the metadata entirely, so there’s no need to use JSON.stringify. and for  Array.includes, I’ll use Set(), as it offers O(1) time complexity, which is better. yup , i need to fix the linting issues.

**Comment by jayantbh** [comment]
Directionally I think yes.
Your last commit looks better too.

I haven't done a proper review but I'll ask the team to check it out.

**Comment by jayantbh** [comment]
This PR could use some tests. Especially with various kinds of logs that might come through.

**Comment by jayantbh** [comment]
And of course, to test the performance of how many lines you're able to process via your functions as a benchmark.

**Comment by jayantbh** [comment]
I ran a quick test myself based on the number of lines, and I think we'll hit other performance limits before we hit perf issues with the regexes and matches.

**Comment by vishalmishraa** [comment]
> I ran a quick test myself based on the number of lines, and I think we'll hit other performance limits before we hit perf issues with the regexes and matches.

okay , i think i need to change my approach ,Could you suggest an alternatives or any way to do that ? I will work on implementing it accordingly.

**Comment by jayantbh** [comment]
No no. I'm saying you might not need to change your approach.

**Comment by vishalmishraa** [comment]
so , do i need to do any improvements or it is good to go ?

**Comment by jayantbh** [comment]
> This PR could use some tests. Especially with various kinds of logs that might come through.

This.

**Comment by vishalmishraa** [comment]
Okay, I’m working on it. I’ll be writing tests in ./web-server/src/utils/test, where I’ll create a logFormatter.test.ts file and write the tests. I need some help with generating different types of system logs. Should I generate new logs for testing, or can I use the logs that are being generated right now?

**Comment by vishalmishraa** [comment]
@jayantbh  I need help regarding generating different types of system logs. Should I generate new logs for testing? If yes, how can I generate other different logs? Or is it better to use the logs that are currently being generated?

**Comment by jayantbh** [comment]
I think handle for the variety of logs that are generated today.
But ensure that when and if something unexpected comes in the logs (such as, missing HTTP method in HTTP logs, dates formatted in a different way, etc.) Then it doesn't break the UI.

**Comment by vishalmishraa** [comment]
i think the function I wrote is not feasible for parsing all types of logs. There are several different types of logs, such as system-generated logs (some print statements in backend code), HTTP server error logs, and others.

i did some research on better logging solutions and found a CNCF-based library called [Fluentd](https://www.fluentd.org/), which offers unified logging, better parsing, and improved data collection and observation. It's also lightweight, requiring only 40 to 50 MB of space. I think it could be a good fit for this project.

more details:
- [What is Fluentd?](https://www.fluentd.org/architecture)
- [How Fluentd works (YouTube video)](https://youtu.be/5ofsNyHZwWE?t=257)

**Comment by jayantbh** [comment]
I suspected it could go in this way.
You can go with a simpler approach altogether as well.

**Comment by jayantbh** [comment]
> > I suspected it could go in this way. You can go with a simpler approach altogether as well.
> 
> Then , there will be some logs, which can't be parsed, and they will be displayed as it is without colors. I'll try to add regex for logs that are frequently generated, should i proceed with this ?

Yea, I think you could try that.

Something is better than nothing after all. But make sure you have tests.

**Comment by vishalmishraa** [comment]
> > > I suspected it could go in this way. You can go with a simpler approach altogether as well.
> > 
> > 
> > Then , there will be some logs, which can't be parsed, and they will be displayed as it is without colors. I'll try to add regex for logs that are frequently generated, should i proceed with this ?
> 
> Yea, I think you could try that.
> 
> Something is better than nothing after all. But make sure you have tests.

Thanks, @jayantbh. I've updated the functions and added the tests. Please review them.

**Comment by jayantbh** [comment]
Hmm, not sure I feel great about that approach.
How did you try to add and test the error boundary that it caused a higher level boundary to catch it instead?

**Comment by vishalmishraa** [comment]
Hi, @jayantbh,  can you please check this now , as i've implemented error handling . also now i'm not using another component , combined it in SystemLog.tsx , with error message .

**Comment by vishalmishraa** [comment]
> Hmm, not sure I feel great about that approach. How did you try to add and test the error boundary that it caused a higher level boundary to catch it instead?

i did this : 



```

const SystemLogs = () => {
    return (
        <ErrorBoundary FallbackComponent={SystemLogErrorFallback}>
            // code
        </ErrorBoundary>
    );
};

```

But it didn’t work, and I think it should have worked.

**Comment by vishalmishraa** [comment]
Also i think the approach I'm using now would be work good  because the component which should be  passed into the ErrorBoundary fallback should also receive serviceName and other props as it should render logs in plain text. So, in the future, if we make any improvements to the UI for SystemLogs like (implementing search feature ,etc..), we would need to apply the same modifications to the custom fallback component as well, since it's rendering system logs in plain text. Therefore, it would be better to handle this in the same position .

**Comment by vishalmishraa** [comment]
Hi @jayantbh ,

I’ve implemented the error boundary and apologize for the confusion earlier. You were right , I was testing it incorrectly, which is why it didn’t work. However, I’ve now implemented it properly. I’ve also created a useSystemLog hook for it and separated the component.Please check it now.

**Comment by jayantbh** [comment]
@vishalmishraa could you share screenshots of the error boundary in effect?

**Comment by vishalmishraa** [comment]
> @vishalmishraa could you share screenshots of the error boundary in effect?

### Sure : 


https://github.com/user-attachments/assets/e05ff22a-ad98-474b-892f-105bd3b3eef6

**Comment by jayantbh** [comment]
Looks great!
The linter appears to have failed. Do check what that's about.

Meanwhile, we'll give this a round of final reviews. But I think we're looking good visually/functionally.

**Comment by vishalmishraa** [comment]
i think linter is failing because of workflow 

Screenshot 2024-10-15 at 2 15 56 PM

**Comment by jayantbh** [comment]
Okay, I guess ignore that for now. We'll do the reviews as soon as we're able. :)

File: web-server/src/utils/logFormatter.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  postgresLogRegex,
+  postgresMultiLineLogRegex,
+  dataSyncLogRegex,
+  validLogLevels
+} from '../constants/log-formatter';
+
+export const parseLogLine = (rawLogLine: string): ParsedLog | null => {
+  const generalLogMatch = rawLogLine.match(generalLogRegex);
+  if (generalLogMatch) {
+    const [, timestamp, , logLevel, message] = generalLogMatch;
```
Comment (e-for-eshaan) [inline web-server/src/utils/logFormatter.ts:15]:
```ts
const [_pickAname, timestamp, _pickAnotherName, logLevel, message]
```

File: web-server/src/utils/logFormatter.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  const httpLogMatch = rawLogLine.match(httpLogRegex);
+  if (httpLogMatch) {
+    const [, ip, , , timestamp, request, status, bytes, referer, userAgent] =
+      httpLogMatch;
+    const [method, path] = request.split(' ');
+    return {
+      timestamp,
+      logLevel: 'INFO', // Assuming all HTTP logs are INFO level
```
Comment (e-for-eshaan) [inline web-server/src/utils/logFormatter.ts:30]: all these string constants should be wrapped in an enum in our `src/types/resources.ts`
```ts
enum YourEnumName 
{ 
  INFO = 'INFO', 
  DEBUG = 'DEBUG'
  ...other_possible_values... 
}
```

File: web-server/src/utils/logFormatter.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      logLevel: validLogLevels.has(normalizedLogLevel)
+        ? normalizedLogLevel
+        : 'INFO',
+      message: message.trim()
+    };
+  }
+
+  const postgresLogMatch = rawLogLine.match(postgresLogRegex);
+  if (postgresLogMatch) {
+    const [, timestamp, , logLevel, message] = postgresLogMatch;
```
Comment (e-for-eshaan) [inline web-server/src/utils/logFormatter.ts:80]: same:
```ts
const [_pickAname, timestamp, _pickAnotherName, logLevel, message]
```

File: web-server/src/utils/logFormatter.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      logLevel: validLogLevels.has(normalizedLogLevel)
+        ? normalizedLogLevel
+        : 'INFO',
+      message
+    };
+  }
+
+  const dataSyncLogMatch = rawLogLine.match(dataSyncLogRegex);
+  if (dataSyncLogMatch) {
+    const [, logLevel, action, service, message] = dataSyncLogMatch;
```
Comment (e-for-eshaan) [inline web-server/src/utils/logFormatter.ts:94]: same

File: web-server/src/constants/log-formatter.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  'INFO',
+  'NOTICE',
+  'WARNING',
+  'ERROR',
+  'LOG',
+  'FATAL',
+  'PANIC',
+  'STATEMENT',
+  'DETAIL'
+]);
```
Comment (e-for-eshaan) [inline web-server/src/constants/log-formatter.ts:31]: this should be an enum in our `resources` file

**Review by e-for-eshaan — CHANGES_REQUESTED** [review]

File: web-server/src/utils/logFormatter.ts (hunk 1 above)
Comment (vishalmishraa) [reply web-server/src/utils/logFormatter.ts:15]: got it.

File: web-server/src/utils/logFormatter.ts (hunk 2 above)
Comment (vishalmishraa) [reply web-server/src/utils/logFormatter.ts:30]: done , created a enum LogLevel.

File: web-server/src/constants/log-formatter.ts (hunk 5 above)
Comment (vishalmishraa) [reply web-server/src/constants/log-formatter.ts:31]: I added these as enums and also removed the validLogLevels logic from the parser. Previously, I was using two approaches to parse PostgreSQL logs, but during testing, I noticed that my first approach, PostgresMultilineLogMatch, can parse both types of PostgreSQL logs. Therefore, there's no need to write separate logic for each type ,refer to the [official postgres documentation  ](https://www.postgresql.org/docs/current/runtime-config-logging.html#:~:text=server%20command%20line.-,log_line_prefix,-(string)). and  [this article ](https://betterstack.com/community/guides/logging/how-to-start-logging-with-postgresql/#structured-logging-in-json:~:text=%25m%20%5B%25p%5D%20%25q,the%20following%20details%3A) also for  [log levels ](https://www.postgresql.org/docs/current/runtime-config-logging.html#:~:text=Table%C2%A019.2.-,Message%20Severity%20Levels,-Severity).

File: web-server/src/utils/logFormatter.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  httpLogRegex,
+  redisLogRegex,
+  postgresLogRegex,
+  dataSyncLogRegex
+} from '../constants/log-formatter';
+
+export const parseLogLine = (rawLogLine: string): ParsedLog | null => {
+  const generalLogMatch = rawLogLine.match(generalLogRegex);
+  if (generalLogMatch) {
+    const [_fullLog, timestamp, _unused, logLevel, message] = generalLogMatch;
```
Comment (e-for-eshaan) [inline web-server/src/utils/logFormatter.ts:15]: `_unused`
this looks off, name it correctly to what this variable stands for, and the underscore is enough to know that this variable is unused

File: web-server/src/utils/logFormatter.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    };
+  }
+
+  const httpLogMatch = rawLogLine.match(httpLogRegex);
+  if (httpLogMatch) {
+    const [
+      _fullLog,
+      ip,
+      _unused,
+      _unused2,
```
Comment (e-for-eshaan) [inline web-server/src/utils/logFormatter.ts:29]: same with these :)

**Review by e-for-eshaan — CHANGES_REQUESTED** [review]

File: web-server/src/utils/logFormatter.ts (hunk 7 above)
Comment (vishalmishraa) [reply web-server/src/utils/logFormatter.ts:29]: changed the second one to _username , and first one with _ (underscore) [refer to this](https://docs.gunicorn.org/en/stable/settings.html#:~:text=remote%20address-,l,%27%2D%27,-u)

File: web-server/src/utils/logFormatter.ts (hunk 6 above)
Comment (vishalmishraa) [reply web-server/src/utils/logFormatter.ts:15]: done. ,changing it with _pid

**Review by e-for-eshaan — DISMISSED** [review]
LGTM, good stuff @vishalmishraa 🎉

File: web-server/src/components/Service/SystemLog/FormattedLog.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    [theme]
+  );
+
+  const { timestamp, ip, logLevel, message } = log;
+  return (
+    <Line
+      key={index}
+      marginBottom={'8px'}
+      fontSize={'14px'}
+      fontFamily={'monospace'}
```
Comment (shivam-bit) [inline web-server/src/components/Service/SystemLog/FormattedLog.tsx:47]: same as other requested changes

File: web-server/src/components/Service/SystemLog/FormattedLog.tsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    >
+      <Line component="span" color="info.main">
```
Comment (shivam-bit) [inline web-server/src/components/Service/SystemLog/FormattedLog.tsx:49]: i think you can do this like `<Line  color="info">`

**Review by shivam-bit — CHANGES_REQUESTED** [review]

File: web-server/src/constants/log-formatter.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+export interface ParsedLog {
+  timestamp: string;
+  logLevel: string;
+  message: string;
+  role?: string;
+  ip?: string;
+}
```
[the remark this answers is not in the document]
Comment (vishalmishraa) [reply web-server/src/constants/log-formatter.ts:7]: done

File: web-server/src/components/Service/SystemLog/FormattedLog.tsx [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      };
+
+      return colors[level.toUpperCase()] || theme.colors.info.main;
+    },
+    [theme]
+  );
+
+  const { timestamp, ip, logLevel, message } = log;
+  return (
+    <Line medium marginBottom={'8px'} fontFamily={'monospace'}>
```
Comment (shivam-bit) [inline web-server/src/components/Service/SystemLog/FormattedLog.tsx:37]: Try this`<Line mono marginBottom={1}>`

File: web-server/src/components/Service/SystemLog/PlainLog.tsx [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import { Line } from '@/components/Text';
+
+export const PlainLog = ({ log }: { log: string; index: number }) => {
+  return (
+    <Line medium marginBottom={'8px'} fontFamily={'monospace'}>
```
Comment (shivam-bit) [inline web-server/src/components/Service/SystemLog/PlainLog.tsx:5]: Try this `<Line mono marginBottom={1}>`, it should ideally give you the same effect.

**Review by shivam-bit — CHANGES_REQUESTED** [review]

File: web-server/src/components/Service/SystemLog/PlainLog.tsx (hunk 12 above)
Comment (vishalmishraa) [reply web-server/src/components/Service/SystemLog/PlainLog.tsx]: done

**Comment by shivam-bit** [comment]
@vishalmishraa LGTM!🚀🚀🚀

**Review by shivam-bit — APPROVED** [review]

**Comment by jayantbh** [comment]
I think this doesn't need another round of reviews. We can go ahead and merge this. Any remaining changes will probably be minor and can be taken up in subsequent PRs.

**Comment by jayantbh** [comment]
Congrats on getting this merged!
Good work my dude!
